File: Lab/SEM4/MPP/lab5/moto_java/src/main/java/org/zdroba/repository/RaceEventRepositoryImpl.java
package org.zdroba.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.zdroba.entity.RaceEvent;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class RaceEventRepositoryImpl implements RaceEventRepository {

    private static RaceEventRepositoryImpl instance = null;

    private RaceEventRepositoryImpl() {}

    public static RaceEventRepositoryImpl getInstance() {
        if (instance == null)
            instance = new RaceEventRepositoryImpl();
        return instance;
    }

    private final Connection connection = DataBaseConnection.getInstance().getConnection();
    private static final Logger logger = LogManager.getLogger();

    @Override
    public RaceEvent find(Long key) {
        logger.traceEntry();
        String sql = "SELECT * FROM events WHERE id = ? LIMIT 1";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setLong(1, key);

            try(ResultSet rs = ps.executeQuery()){
                if(rs.next()){
                    int engine = rs.getInt("engine");

                    RaceEvent event = new RaceEvent(engine);
                    event.setId(key);

                    logger.traceExit();
                    return event;
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return null;
    }

    @Override
    public List<RaceEvent> getAll() {
        logger.traceEntry();
        String sql = "SELECT * FROM events";

        List<RaceEvent> events = new ArrayList<>();

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            try(ResultSet rs = ps.executeQuery()){

                while(rs.next()){
                    Long id = rs.getLong("id");
                    int engine = rs.getInt("engine");

                    RaceEvent event = new RaceEvent(engine);
                    event.setId(id);

                    events.add(event);
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return events;
    }

    @Override
    public void add(RaceEvent raceEvent) {
        logger.traceEntry("saving RaceEvent{}", raceEvent);
        String sql = "INSERT INTO events (engine) VALUES (?)";

        try(PreparedStatement ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)){
            ps.setInt(1, raceEvent.getEngine());

            int result = ps.executeUpdate();

            ResultSet rs = ps.getGeneratedKeys();
            if (rs.next()) {
                raceEvent.setId(rs.getLong(1));
            }

            logger.trace("Saved {} instance ", result);
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public RaceEvent find(int engine) {
        logger.traceEntry();
        String sql = "SELECT * FROM events WHERE engine = ? LIMIT 1";

        try(PreparedStatement ps = connection.prepareStatement(sql)){
            ps.setInt(1, engine);

            try(ResultSet rs = ps.executeQuery()){

                if(rs.next()){
                    Long id = rs.getLong("id");

                    RaceEvent event = new RaceEvent(engine);
                    event.setId(id);

                    logger.traceExit();
                    return event;
                }
            }
        } catch (SQLException e) {
            logger.error(e.getMessage());
            System.err.println(e.getMessage());
        }

        logger.traceExit();
        return null;
    }
}

File: Lab/SEM4/MPP/lab4/moto_c/repository/RaceEventRepositoryImpl.cs
using moto_c.entity;
using MySqlConnector;
using NLog;

namespace moto_c.repository;

public class RaceEventRepositoryImpl:RaceEventRepository
{
    private readonly MySqlConnection _connection = DatabaseConnection.getInstance().GetConnection();
    private static readonly Logger logger = LogManager.GetCurrentClassLogger();
    
    private static RaceEventRepositoryImpl? instance;

    private RaceEventRepositoryImpl() {}

    public static RaceEventRepositoryImpl getInstance()
    {
        if (instance == null)
            instance = new RaceEventRepositoryImpl();
        return instance;
    }
    
    public List<RaceEvent> getAll()
    {
        logger.Trace("Entry GetAll");
        string sql = "SELECT * FROM events";
        List<RaceEvent> events = new List<RaceEvent>();
        
        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    while (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        int engine = rs.GetInt32("engine");
                        
                        RaceEvent raceEvent = new RaceEvent(engine);
                        raceEvent.id = id;
                        
                        events.Add(raceEvent);
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }
        
        logger.Trace("Exit GetAll");
        return events;
    }

    public void add(RaceEvent raceEvent)
    {
        logger.Trace("Entry Add with engine={}", raceEvent.engine);
        string sql = "INSERT INTO events (engine) VALUES (@engine)";

        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@engine", raceEvent.engine);
                int result = cmd.ExecuteNonQuery();
                logger.Trace("Saved {} instance", result);
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }
        
        logger.Trace("Exit Add");
    }

    public RaceEvent? find(int engine)
    {
        logger.Trace("Entry Find with engine={}", engine);
        string sql = "SELECT * FROM events WHERE engine=@engine";
        
        try
        {
            using (MySqlCommand cmd = new MySqlCommand(sql, _connection))
            {
                cmd.Parameters.AddWithValue("@engine", engine);

                using (MySqlDataReader rs = cmd.ExecuteReader())
                {
                    if (rs.Read())
                    {
                        long id = rs.GetInt64("id");
                        
                        RaceEvent raceEvent = new RaceEvent(engine);
                        raceEvent.id = id;
                        
                        logger.Trace("Exit Find with engine={}", engine);
                        return raceEvent;
                    }
                }
            }
        }
        catch (MySqlException e)
        {
            logger.Error(e.Message);
            Console.Error.WriteLine(e.Message);
        }
        
        logger.Trace("Exit Find with engine={}", engine);
        return null;
    }
}